sdk/storage/Azure.Storage.Common/tests/Shared/ClientBuilder.cs

```diff
@@ -0,0 +1,178 @@
+        private readonly GetServiceClientOptions _getServiceClientOptions;
+
+        private TenantConfigurationBuilder Tenants { get; }
+
+        public RecordedTestBase AzureCoreRecordedTestBase => Tenants.AzureCoreRecordedTestBase;
+
+        public TestRecording Recording => AzureCoreRecordedTestBase.Recording;
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="tenantConfigurationBuilder">Source of tenants for appropriate service clients.</param>
+        /// <param name="getServiceClient">Constructs a service client from the given arguments.</param>
+        /// <param name="getServiceClientStorageSharedKeyCredential">Constructs a service client from the given arguments.</param>
+        /// <param name="getServiceClientTokenCredential">Constructs a service client from the given arguments.</param>
+        /// <param name="getServiceClientAzureSasCredential">Constructs a service client from the given arguments.</param>
+        /// <param name="instrumentClient">Delegate to the test environment's InstrumentClient functionality.</param>
+        public ClientBuilder(
+            TenantConfigurationBuilder tenantConfigurationBuilder,
+            GetServiceClient getServiceClient,
+            GetServiceClientStorageSharedKeyCredential getServiceClientStorageSharedKeyCredential,
+            GetServiceClientTokenCredential getServiceClientTokenCredential,
+            GetServiceClientAzureSasCredential getServiceClientAzureSasCredential,
+            GetServiceClientOptions getServiceClientOptions)
+        {
+            Tenants = tenantConfigurationBuilder;
+            _getServiceClient = getServiceClient;
+            _getServiceClientStorageSharedKeyCredential = getServiceClientStorageSharedKeyCredential;
+            _getServiceClientTokenCredential = getServiceClientTokenCredential;
+            _getServiceClientAzureSasCredential = getServiceClientAzureSasCredential;
+            _getServiceClientOptions = getServiceClientOptions;
+        }
+
+        public TServiceClient GetServiceClient_SharedKey(TServiceClientOptions options = default)
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigDefault, options);
+
+        public TServiceClient GetServiceClient_SecondaryAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigSecondary);
+
+        public TServiceClient GetServiceClient_PreviewAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPreviewBlob);
+
+        public TServiceClient GetServiceClient_PremiumBlobAccount_SharedKey()
+            => GetServiceClientFromSharedKeyConfig(Tenants.TestConfigPremiumBlob);
+        public TServiceClient GetServiceClient_OAuth() =>
+            GetServiceClientFromOauthConfig(Tenants.TestConfigOAuth);
+
+        public TServiceClient GetServiceClient_OAuthAccount_SharedKey() =>
+            GetServiceClientFromSharedKeyConfig(Tenants.TestConfigOAuth);
```

Comment:
for the future perhaps. I think that separate "Oauth" tenant doesn't make sense, credentials is orthogonal concept to account feature set , i.e. I can connect with token credential to all of these account here (except files).
I.e. default tenant and oauth tenant can be the same thing.

Reply:
Definitely a good idea but that is a huge scope creep for this PR.

Reply:
can you log work item for this?

Reply:
#24458